### ISSUE TITLE
500 errors on APIs with "FATAL:  remaining connection slots are reserved for non-replication superuser connections"

### ISSUE THREAD

**Issue by bhgrant8** [hackoregon/civic-devops#178]
Teams across projects are running into an error with is resulting in 500s from APIs across project teams. This is causing downtime for development. 

Running the transportation systems project locally with the `-p` flag  and debug as true gives me the following error, which seems to point to an issue with postgres connections:

```
FATAL:  remaining connection slots are reserved for non-replication superuser connections
```

This seems to be related to the question of idle database connections:
https://github.com/hackoregon/civic-devops/issues/177

Transportation systems attempted a fix in the above issue, but does not seem to necesarrily resolve.

I am thinking the connections limit we are running into is more related to this:
https://www.postgresql.org/docs/current/static/runtime-config-connection.html#GUC-MAX-CONNECTIONS then the max limits on the django apis? Are there 100 connections currently open overall?

So it might be possible to increase our max limit on the postgres server, if we are not running into ram issues?

That said the root of #177 still needs further investigation as we do not want to keep running up until limit and either manually close or increase

**Comment by MikeTheCanuck** [comment]
The server is definitely approaching the max_connections = 100 limit:

<img width="563" alt="screen shot 2018-06-17 at 22 36 07" src="https://user-images.githubusercontent.com/6953053/41519895-ae1ca18a-727f-11e8-8bf7-01a19bfcdc0a.png">

I've peeked at the server throughout the day and the number of connections steadily climbs after every attempt to close idle connections.

**Comment by MikeTheCanuck** [comment]
Under every other circumstance leading up to this weekend, the number of idle connections on the db server was never a noticeable issue.

Now we're experiencing this, and while it's entirely possible to up the `max_connections` limit, that won't solve the problem of unbounded growth in number of long-lived idle connections - and it *will* eventually exhaust available memory on the server to no one's real benefit.

Brian, the fix you've implemented on the TS API has definitely provided some relief - even at this ~100-connection zone, the TS database only has 10 connections, whereas LE currently has 14 and ND currently has 50.

I'm recommending some form of connection pooling be implemented - is that what you did in [PR 71](https://github.com/hackoregon/transportation-systems-backend-2018/pull/71) or [PR 72](https://github.com/hackoregon/transportation-systems-backend-2018/pull/72) @bhgrant8, or would that be in addition?

**Comment by bhgrant8** [comment]
@MikeTheCanuck yes changes made would be in 71 and 72. All changes were in the gunicorn_conf.py. This file should be standard across projects, so could be copied over. My main question before we say this has added relief is whether this drop in connections is related to this change vs. that trans-sys has been less active in the last few hours?

i have jmeter running on my local machine now, and in process of getting the prod container to run against a local db container. I want to basically run a test as follows:

1. wire up the prod docker-compose to run a local db container
2. run the prod api container locally, 
3. check pgadmin connections prior to running load tests
4. run load tests (at least 100 on connection points)
5. check pgadmin for active connections after

**Comment by MikeTheCanuck** [comment]
Trying to decipher all the options in front of us, it's quite a puzzle.  What I'm veering towards:
- closing connections more aggressively is a good thing no matter what, but is only a partial mitigation and can still suffer from peak pile-ups that soak up all available connections at the DB
- reducing the max number of connections from each Django source is a good thing, but as we continue to add new containers this still inevitably will hit the wall pretty quickly e.g. 10 APIs x 2 containers each (for robustness) with 5 max connections each = the 100 connections limit
- pooling requests to a single longer-lived connection feels the most correct way of scaling our infrastructure - not only because each Django app can be firing a ton of requests and the setup/teardown of TCP sessions is just a waste, but also because from the db's perspective there's no point in holding 10 connections open to a single request source when 1 would do just as well.   However, from my initial survey of the space, I'm getting the impression that the interactions between Django, gunicorn, gevent and a load-balancer aren't trivial to reason out, and we'll need to experiment a little to figure out an architecture that actually achieves our goal.

**Comment by MikeTheCanuck** [comment]
Closing connections using Brian's strategy from above:
- disaster-resilience-backend, elections-2018-backend, housing-2018 projects needed the update
- neighborhoods-2018 project didn't need the update
